[choongzhanhong] iP

choongzhanhong opened the pull request:
Add updates up to Level-3
Implement commands to add task, mark task done/undone, list tasks.
Add Message class for output strings and help messages.

choongzhanhong committed:
.gitignore: add *.class
Duke.java: add level-0 greet and exit message

choongzhanhong committed:
Add input and echo

PAPA (Previously named Duke) now waits for input and will echo the
input.
Greeting messages and miscellaneous strings put in separate
Messages class.

choongzhanhong committed:
Add task list functionality

The user can now add up to 100 tasks to the list through input.
Task list can be listed out using "list" command.

choongzhanhong committed:
Add mark as done/undone feature

You can now mark items on the task list as done/undone using
"mark #" and "unmark #", # being the item number on the list.

choongzhanhong committed:
Add "help" command

Commit made so as to sync files across laptop and home desktop.

choongzhanhong committed:
Message: refactor constants

Renamed and re-indented strings to better fit coding style.
Some minor tweaks to printed messages as well.

choongzhanhong committed:
Add Todo, Deadline, Event classes

Added the three subclasses and minor tweaks to Task class.

choongzhanhong committed:
Refactor Message to Command

The Command class will now be used to hold methods to parse input.
It will continue to hold output text as well.

choongzhanhong committed:
Update Duke.java

Changed the task function to accept the Todo, Deadline, and Event.
No input validation at this point.
Code compiles but is incomplete. Pushing to work on other computer.

choongzhanhong committed:
Update Duke.java

Now accepts Todo, Deadline, and Event inputs.
Does not validate input yet.
Code is still a little messy.

ysl-28 commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,116 @@
+    private static final String COMMAND_TASK_EVENT = "event";
+    private static final String COMMAND_LIST = "list";
+    private static final String COMMAND_BYE = "bye";
+    private static final String COMMAND_MARK = "mark";
+    private static final String COMMAND_UNMARK = "unmark";
+
+    public static final String MESSAGE_HELP = "PAPA is your personal task tracker. Options:\n" +
+            "Add tasks: \n" +
+            String.format("   %s <task_name>\n", COMMAND_TASK_TODO) +
+            String.format("   %s <task_name> /by <due_date>\n", COMMAND_TASK_DEADLINE) +
+            String.format("   %s <task_name> /from <start> /to <end>\n", COMMAND_TASK_EVENT) +
+            String.format("%s        Show these tips.\n", COMMAND_HELP) +
+            String.format("%s        List out existing tasks.\n", COMMAND_LIST) +
+            String.format("%s         Exit PAPA.\n", COMMAND_BYE) +
+            String.format("%s <n>    Mark the n-th task as done.\n", COMMAND_MARK) +
+            String.format("%s <n>  Mark the n-th task as undone.\n", COMMAND_UNMARK);
+
+    public static final String MESSAGE_LOGO =
+            "██████╗  █████╗ ██████╗  █████╗ \n" +
+            "██╔══██╗██╔══██╗██╔══██╗██╔══██╗\n" +
+            "██████╔╝███████║██████╔╝███████║\n" +
+            "██╔═══╝ ██╔══██║██╔═══╝ ██╔══██║\n" +
+            "██║     ██║  ██║██║     ██║  ██║\n" +
+            "╚═╝     ╚═╝  ╚═╝╚═╝     ╚═╝  ╚═╝";
+    public static final String MESSAGE_INTRO =
+            "Hello! I'm PAPA, your Personal Assistant, Personal Angel.\n" +
+            "What can I do for you? Type 'help' for a list of commands.";
+    public static final String MESSAGE_OUTRO = "Bye. Hope to see you again soon!";
+    private static final String MESSAGE_TASK_ADDED =
+            "Great! I've added the task for you:";
+    private static final String MESSAGE_TASK_COUNT =
+            "Now you have %d tasks"; // Format at runtime.
+    public static final String MESSAGE_TASKS_EMPTY =
+            "Looks like you don't have anything to do. Nice!";
+    public static final String MESSAGE_TASK_DONE =
+            "Great! This task is now done: ";
+    public static final String MESSAGE_TASK_UNDONE =
+            "Okay, I've marked this task as not done yet: ";
+    private static final String MESSAGE_INVALID_COMMAND =
+            "Oops, I don't understand you. Type 'help' for commands you can use!";
```
Good naming style for constant names.

ysl-28 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,94 @@
+import java.util.Scanner;
+
 public class Duke {
+    private static final TaskList tasks = new TaskList();
```
Good use of plural form to name a collection of objects.

ysl-28 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,94 @@
+        }
+
+        // Check user input against commands.
+        if (input.equals("bye")) {
+            return;
+        } else if (input.equals("list")) {
+            System.out.print(tasks.getTaskListString());
+        } else if (input.equals("help")) {
+            System.out.print(Command.MESSAGE_HELP);
+        } else {
+            String[] commands = Command.splitCommandAndArgs(input);
+            addNewTask(commands[0], commands[1]);
+        }
```
Correct form for if-else statement.

Geeeetyx replied:
Correct form for conditionals on another line for if-else statement

ysl-28 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,94 @@
+     */
+    private static void addNewTask(String command, String commandArgs) {
+        Task newTask;
+        switch(command) {
+        case "todo":
+            newTask = new Todo(commandArgs, TaskList.getNextTaskNumber());
+            //args should just be the Description
+            tasks.addTask(newTask);
+            Command.printTaskAdded(newTask);
+            break;
```
Correct form for writing switch statement.

ysl-28 commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.Arrays;
+
+public class TaskList {
```
Correct use of PascalCase to name classes.

ysl-28 reviewed (commented):
Overall, the code follows coding standards and makes good use of constants.

SebastianSoewanto commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,116 @@
+            String.format("%s        List out existing tasks.\n", COMMAND_LIST) +
+            String.format("%s         Exit PAPA.\n", COMMAND_BYE) +
+            String.format("%s <n>    Mark the n-th task as done.\n", COMMAND_MARK) +
+            String.format("%s <n>  Mark the n-th task as undone.\n", COMMAND_UNMARK);
```
Very neat format and naming style

SebastianSoewanto commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,116 @@
+    /**
+     * Task added
+     */
+    public static void printTaskAdded(Task task) {
+        System.out.println(MESSAGE_TASK_ADDED);
+        System.out.println(task);
+        System.out.println("Total number of tasks: " + TaskList.getNumberOfTasks());
+    }
+
+    /**
+     * Split input into 1st arg and subsequent line.
+     * Taken from Lecture - Contacts (Wk 4 content) as it is relevant.
+     * @param inputLine A line of input.
+     */
+    public static String[] splitCommandAndArgs(String inputLine) {
+        // Trim whitespace, then split once. Output is [command, args].
+        final String[] split = inputLine.trim().split("\\s+", 2);
+        // Else case: no parameters
+        return split.length == 2 ? split : new String[]{split[0], ""};
+    }
```
very concise use of ternary operator

Geeeetyx replied:
Good naming of methods in camelCase

SebastianSoewanto commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,25 @@
+        String eventTime = String.format(" from %s to %s", this.timeStart, this.timeEnd);
+        return "[E]" + super.toString() + eventTime;
+    }
+}
```
Good implementation of class

Geeeetyx commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,116 @@
+    /**
+     * Inputs go here
+     */
+    private static final String COMMAND_HELP = "help";
+    private static final String COMMAND_TASK_TODO = "todo";
+    private static final String COMMAND_TASK_DEADLINE = "deadline";
+    private static final String COMMAND_TASK_EVENT = "event";
+    private static final String COMMAND_LIST = "list";
+    private static final String COMMAND_BYE = "bye";
+    private static final String COMMAND_MARK = "mark";
+    private static final String COMMAND_UNMARK = "unmark";
```
Good constant names that follow all uppercase with underscore to seperate words

Geeeetyx commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,19 @@
+public class Deadline extends Task {
+    protected String by;
+
+    /**
+     * A task with a deadline.
+     * @param description Name of task
+     * @param by Deadline
+     */
```
Correct form for comments

SebastianSoewanto commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,40 @@
+
+    // return a string to indicate done or not done.
+    private String getDoneIcon() {
+        return (this.isDone ? "[X]" : "[ ]"); // mark done task with X
```
Suggest to include the comment together above to explain the function together

Geeeetyx commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,40 @@
+        return (this.getDoneIcon() + " " + this.description);
+    }
+
+    // return a string to indicate done or not done.
+    private String getDoneIcon() {
+        return (this.isDone ? "[X]" : "[ ]"); // mark done task with X
+    }
```
Can consider using if-else statement for more clarity

SebastianSoewanto reviewed (commented):
Good job on the overall organization of your code!

Geeeetyx reviewed (commented):
Good job overall, the change in name for the program to PAPA was very creative

choongzhanhong committed:
Add text ui test

choongzhanhong committed:
Add input validation

So far only added for mark/unmark. The tasks are WIP.
Currently works, but a little messy.

choongzhanhong committed:
Add input validation for Task

Now validates the 3 task inputs, but not exhaustively.
Also refactor Command again to IO (Handling input/output)

choongzhanhong committed:
Merge branch 'branch-Level-5'

choongzhanhong committed:
Merge branch 'master' of https://github.com/choongzhanhong/ip

choongzhanhong committed:
Package tasks and io

Two separate packages, one for task related things, one for i/o.

choongzhanhong committed:
Merge branch 'A-Packages'

kohnh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,110 @@
+     * ==============================================================
+     * Below handling Tasks (Validation and addTask)
+     * ==============================================================
+     */
```
This seems like an indicator for yourself. Maybe you should create individual JavaDocs for each of your methods for easy reference instead of organizing by sections

kohnh commented on src/main/java/io/DukeException.java:
```diff
@@ -0,0 +1,5 @@
+package io;
+
+public class DukeException extends Exception{
+    // Nothing needed here
```
The comments given are not really adding any value and as such should be removed.

kohnh commented on src/main/java/io/IO.java:
```diff
@@ -0,0 +1,172 @@
+
+/*
+* Use this class for INPUT and OUTPUT (IO) related stuff
+* Got some inspiration from Contacts (Week 4)
```
Similar to the above comments that should be in the JavaDocs form which informs what the class is about or the functionality but good that you credited where you got your inspiration from (maybe you can specify more if not outside readers may be a little confused)

kohnh commented on src/main/java/io/IO.java:
```diff
@@ -0,0 +1,172 @@
+    public static String[] splitCommandAndArgs(String inputLine) {
+        // Trim trailing and beginning whitespace, then split once. Output is [command, args].
+        final String[] split = inputLine.trim().split("\\s+", 2);
+        // Else case: no parameters
```
```suggestion
        // Edge case: no parameters
```
But you can also consider raising an exception

kohnh commented on src/main/java/io/IO.java:
```diff
@@ -0,0 +1,172 @@
+
+        // Needs to be specifically 2, i.e. [task_name, deadline]
+        // Is this a magic number?
+        if (deadlineArgs.length != 2) {
```
Yes this is considered a magic number. You should consider using a named constant to help the reader understand.
Similarly remove comments that are targeted towards yourself.

kohnh commented on src/main/java/io/IO.java:
```diff
@@ -0,0 +1,172 @@
+     * Note string format placeholders, so ensure to use printf/format.
+     * ==============================================================
+     */
+    public static final String MESSAGE_HELP = "PAPA is your personal task tracker. Options:\n" +
+            "Add tasks: \n" +
```
```suggestion
    public static final String MESSAGE_HELP = "PAPA is your personal task tracker. Options:\n"
            + "Add tasks: \n"
```
Should try to put the operators in front so that readers do not need to keep referring to the previous line

kohnh commented on src/main/java/io/IO.java:
```diff
@@ -0,0 +1,172 @@
+            "Type 'task' to list out the existing tasks!";
+
+
+    // Prints a horizontal line of 32 '=' characters.
```
This comment literally explains what the method does which does not add value

kohnh commented on src/main/java/task/Event.java:
```diff
@@ -0,0 +1,27 @@
+    /**
+     * Constructor for Event
+     * Task with start and end time.
+     * @param description
+     * @param taskNumber
+     * @param timeStart
+     * @param timeEnd
```
If you created the JavaDocs you should try not to leave it blank

kohnh commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,67 @@
+import java.util.Arrays;
+
+public class TaskList {
+    private static final Task[] tasks = new Task[100];
```
Try not to use magic numbers like `100`, instead use a named constant.

kohnh commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    // use "mark" to mark done
+    public String markAsDone() {
+        this.isDone = true;
+        return "Marked as done: " + this;
+    }
+    // use "unmark" to mark a task not done
+    public String markAsUndone() {
+        this.isDone = false;
+        return "Marked as undone: " + this;
+    }
```
Maybe you can consider finding a way to abstract this method since it is pretty similar like what you mentioned for `executeMarkUnmark` method

kohnh commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    // use "mark" to mark done
+    public String markAsDone() {
```
```suggestion
    public String setDone() {
```
Naming should be more as a setter since it is modifying the variables of the class

kohnh reviewed (commented):
Overall, good job using `public static final` to reduce the string literals. Just take note to review the comments that you added

choongzhanhong committed:
Add 'delete' functionality

Also changed underlying data structure to ArrayList.

choongzhanhong committed:
Add file write feature

File reading feature still underway.

choongzhanhong committed:
Add read file feature

Currently somewhat incomplete

choongzhanhong committed:
Add read/write feature

choongzhanhong committed:
Merge branch 'branch-Level-6'

choongzhanhong committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/task/Task.java
#	src/main/java/task/TaskList.java

choongzhanhong committed:
Update gitignore

Ignore output file.

choongzhanhong committed:
Add Manifest

choongzhanhong committed:
Add Ui Class

Extracted from IO. This class will handle user interactions.

choongzhanhong committed:
Add Storage Class

Initial commit. Extracted from IO. For better OOP.

choongzhanhong committed:
Update Storage class

Redesigned Open and Load methods.

choongzhanhong committed:
Merge branch 'oop-extract-storage'

choongzhanhong committed:
Merge branch 'master' into oop-extract-ui

choongzhanhong committed:
Merge branch 'oop-extract-ui'

choongzhanhong committed:
Refactor Ui

I realised you can just use the built-in extract methods to
refactor instead of doing it manually. This commit is
a lesson learned.

choongzhanhong committed:
Merge branch 'oop-extract-ui'

choongzhanhong committed:
Refactor Storage

Redirecting main to use Storage class methods instead of IO.

choongzhanhong committed:
Refactor Storage

Duke main now using Storage instead of IO.

choongzhanhong committed:
Merge branch 'refactor-storage'

choongzhanhong committed:
Add Parser Class

Methods in IO refactored to Parser.
IO Safely deleted.

choongzhanhong committed:
Extract methods to Parser

Moved some parsing methods into Parser.

choongzhanhong committed:
Merge branch 'oop-Parser'

choongzhanhong committed:
Add load method from Storage to TaskList.

This is called in Duke.main() to retrieve tasks from storage.

choongzhanhong committed:
Remove static modifier

TaskList is now non-static. Should have done this long ago.

choongzhanhong committed:
Redesign TaskList

Not sure why I even put TaskNumber in Task instead of in TaskList.
Took that out.

choongzhanhong committed:
Merge branch 'oop-TaskList'

choongzhanhong committed:
Move execute methods to Parser

Duke is now cleaner. It compiles and runs!

choongzhanhong committed:
Add find feature

choongzhanhong committed:
Merge pull request #1 from choongzhanhong/branch-Level-9

Add find feature

choongzhanhong committed:
Add more Javadocs

choongzhanhong committed:
Add UserGuide

choongzhanhong committed:
Create _config.yml
